File: Examples/Objective-C/Navigine Example/ViewController.m
//
//  ViewController.m
//  Navigine Example
//
//  Created by Pavel Tychinin on 19/07/2017.
//  Copyright © 2017 Navigine. All rights reserved.
//

#import "ViewController.h"

@interface ViewController (){
    UIBezierPath   *uipath;
    CAShapeLayer   *routeLayer;
}
@property (weak, nonatomic) IBOutlet UIButton *btnIncreaseFloor;
@property (weak, nonatomic) IBOutlet UIButton *btnDecreaseFloor;
@property (weak, nonatomic) IBOutlet UILabel *lblCurrentFloor;

@property (weak, nonatomic) IBOutlet NSLayoutConstraint *imageViewHeightConstraint;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *imageViewWidthConstraint;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *imageViewTopConstraint;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *imageViewLeadingConstraint;

@property (nonatomic, strong) MapPin *pressedPin;
@property (nonatomic, assign) BOOL isRouting;
@property (nonatomic, strong) NavigineCore *navigineCore;
@property (nonatomic, strong) NCSublocation *sublocation;
@property (nonatomic, assign) NSInteger floor;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
     _floor = 0;
     _isRouting = NO;
    
    _sv.delegate = self;
    _sv.pinchGestureRecognizer.enabled = YES;
    _sv.minimumZoomScale = 1.f;
    _sv.zoomScale = 1.f;
    _sv.maximumZoomScale = 2.f;
    
    _btnIncreaseFloor.hidden = YES;
    _btnDecreaseFloor.hidden = YES;
    _lblCurrentFloor.hidden  = YES;
    _btnDecreaseFloor.transform = CGAffineTransformMakeRotation(M_PI);
    
    [_sv addSubview:_imageView];
    
    _navigineCore = [[NavigineCore alloc] initWithUserHash: @"628B-9792-0789-C136"
                                                    server: @"https://api.navigine.com"];
    _navigineCore.delegate = self;
//    _navigineCore.navigationDelegate = self;
    // Point on map
    _current = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, 30, 30)];
    _current.backgroundColor = [UIColor redColor];
    _current.layer.cornerRadius = _current.frame.size.height/2.f;
    [_imageView addSubview:_current];
    _imageView.userInteractionEnabled = YES;
    
    [NSTimer scheduledTimerWithTimeInterval:0.1f
                                     target:self
                                   selector:@selector(navigationTick:)
                                   userInfo:nil
                                    repeats:YES];

    UITapGestureRecognizer *tapPress = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapPress:)];
    tapPress.delaysTouchesBegan   = NO;
    [_sv addGestureRecognizer:tapPress];
//    [_navigineCore addBeaconGenerator:@"8EEB497E-4928-44C6-9D92-087521A3547C" major:9001 minor:36 timeout:10 rssiMin:-90 rssiMax:-70];
    [_navigineCore downloadLocationById:2475
                            forceReload:true
                           processBlock:^(NSInteger loadProcess) {
                               NSLog(@"%zd",loadProcess);
                           } successBlock:^(NSDictionary *userInfo) {
                               [self setupFloor: _floor];
                               _btnIncreaseFloor.hidden = (_navigineCore.location.sublocations.count == 1);
                               _btnDecreaseFloor.hidden = (_navigineCore.location.sublocations.count == 1);
                               _lblCurrentFloor.hidden  = (_navigineCore.location.sublocations.count == 1);
                           } failBlock:^(NSError *error) {
                               NSLog(@"%@",error);
                           }];
}

- (void) navigationTick: (NSTimer *)timer {
    NCDeviceInfo *res = _navigineCore.deviceInfo;
    if (res.error.code == 0) {
        _current.hidden = res.sublocation != _sublocation.id;
        _current.center = CGPointMake(_imageView.width / _sv.zoomScale * res.kx,
                                      _imageView.height / _sv.zoomScale * (1.0 - res.ky));
    }
    else {
        _current.hidden = YES;
    }
    if (_isRouting) {
        NCRoutePath *devicePath = res.paths.firstObject;
        if (devicePath) {
            NSArray *path = devicePath.points;
            float distance = devicePath.lenght;
            [self drawRouteWithPath:path andDistance:distance];
        }
    }
}

-(void) drawRouteWithPath: (NSArray *)path
              andDistance: (float)distance {
    //    // We check that we are close to the finish point of the route
    if (distance <= 3.) {
        [self stopRoute];
    }
    else {
        [routeLayer removeFromSuperlayer];
        [uipath removeAllPoints];
        
        uipath     = [[UIBezierPath alloc] init];
        routeLayer = [CAShapeLayer layer];
        
        for (int i = 0; i < path.count; i++ ) {
            NCLocationPoint *point = path[i];
            
            if (point.sublocation != _sublocation.id)
                continue;
            
            CGSize imageSizeInMeters = CGSizeMake(_sublocation.width, _sublocation.height);
            
            CGFloat xPoint =  (point.x.doubleValue / imageSizeInMeters.width) * (_imageView.width / _sv.zoomScale);
            CGFloat yPoint =  (1. - point.y.doubleValue / imageSizeInMeters.height)  * (_imageView.height / _sv.zoomScale);
            if (uipath.empty) {
                [uipath moveToPoint:CGPointMake(xPoint, yPoint)];
            }
            else {
                [uipath addLineToPoint:CGPointMake(xPoint, yPoint)];
            }
        }
    }
    routeLayer.hidden = NO;
    routeLayer.path            = [uipath CGPath];
    routeLayer.strokeColor     = [kColorFromHex(0x4AADD4) CGColor];
    routeLayer.lineWidth       = 2.0;
    routeLayer.lineJoin        = kCALineJoinRound;
    routeLayer.fillColor       = [[UIColor clearColor] CGColor];
    
    [_imageView.layer addSublayer:routeLayer];
    [_imageView bringSubviewToFront:_current];
}

- (void)mapPinPressed:(id)sender {
    MapPin *mapPin = (MapPin *)sender;
    [_pressedPin.popUp removeFromSuperview];
    _pressedPin.popUp.hidden = YES;
    
    _pressedPin = mapPin;
    [_imageView addSubview:mapPin.popUp];
    mapPin.popUp.hidden = NO;
    
    mapPin.popUp.bottom   = mapPin.top - 9.0f;
    mapPin.popUp.centerX  = mapPin.centerX;
    
    [mapPin.popUp addTarget:self action:@selector(popUpPressed:) forControlEvents:UIControlEventTouchUpInside];
}

- (void)popUpPressed:(id)sender {
    NCDeviceInfo *res = _navigineCore.deviceInfo;
    
    CGSize imageSizeInMeters = CGSizeMake(_sublocation.width, _sublocation.height);
    CGFloat xPoint = _pressedPin.centerX /_imageView.width * imageSizeInMeters.width;
    CGFloat yPoint = (1. - _pressedPin.centerY /_imageView.height) * imageSizeInMeters.height;
    
    NCLocationPoint *point = [NCLocationPoint pointWithLocation: res.location
                                                    sublocation: _sublocation.id
                                                              x: @(xPoint)
                                                              y: @(yPoint)];
    [_navigineCore cancelTargets];
    [_navigineCore setTarget:point];
    
    [_pressedPin.popUp removeFromSuperview];
    _pressedPin.popUp.hidden = YES;
    _isRouting = YES;
}

- (void)stopRoute {
    _isRouting = NO;
    
    [routeLayer removeFromSuperlayer];
    routeLayer = nil;
    
    [uipath removeAllPoints];
    uipath = nil;
    [_navigineCore cancelTargets];
}

- (void)tapPress:(UITapGestureRecognizer *)gesture {
    [_pressedPin.popUp removeFromSuperview];
    _pressedPin.popUp.hidden = YES;
}

#pragma mark UIScrollViewDelegate methods

- (UIView *)viewForZoomingInScrollView:(UIScrollView *)scrollView{
    return _imageView;
}

#pragma mark NavigineCoreDelegate methods

- (void) didRangePushWithTitle:(NSString *)title
                       content:(NSString *)content
                         image:(NSString *)image
                            id:(NSInteger)id{
    // Your code
}

-(void) setupFloor: (NSInteger) floor {
    [_navigineCore startNavigine];
    [_navigineCore startPushManager];
    
    [_imageView removeAllSubviews];
    _imageView.layer.sublayers = nil;
    [_imageView addSubview:_current];
    
    _sublocation = _navigineCore.location.sublocations[floor];
    
    NSData *imageData = _sublocation.pngImage;
    UIImage *image = [UIImage imageWithData:imageData];
    
    float scale = 1.f;
    if (image.size.width / image.size.height >
        self.view.frame.size.width / self.view.frame.size.height) {
        scale = self.view.frame.size.height / image.size.height;
    }
    else {
        scale = self.view.frame.size.width / image.size.width;
    }
    
    _imageViewWidthConstraint.constant = image.size.width * scale;
    _imageViewHeightConstraint.constant = image.size.height * scale;
    
    _imageViewTopConstraint.constant = 0;
    _imageViewLeadingConstraint.constant = 0;
    
    _imageView.image = image;
    
    _sv.contentSize = CGSizeMake(image.size.width * scale, image.size.height * scale);
    _sv.contentOffset = CGPointMake(-(_imageViewWidthConstraint.constant - self.view.frame.size.width) / 2.0, -(_imageViewHeightConstraint.constant - self.view.frame.size.height) / 2.0);
    _lblCurrentFloor.text = [NSString stringWithFormat:@"%ld", (long)_floor];
    
    [self.view layoutIfNeeded];
    
    [self drawZones];
    [self drawVenues];
}

- (void) drawVenues {
    for (NCVenue *venue in _sublocation.venues) {
        CGFloat xPoint =  venue.x.doubleValue * _imageView.width / _sublocation.width;
        CGFloat yPoint =  _imageView.height * (1 - venue.y.doubleValue / _sublocation.height);
        
        CGPoint point = CGPointMake(xPoint, yPoint);
        MapPin *mapPin = [[MapPin alloc] initWithVenue:venue];
        UIImage *image = [UIImage imageNamed:@"elmVenueIcon"];
        [mapPin setImage:image forState:UIControlStateNormal];
        [mapPin setImage:image forState:UIControlStateHighlighted];
        [mapPin addTarget:self action:@selector(mapPinPressed:) forControlEvents:UIControlEventTouchUpInside];
        [mapPin sizeToFit];
        [_imageView addSubview:mapPin];
        [_sv bringSubviewToFront:mapPin];
        
        mapPin.center  = point;
    }
}

- (void) drawZones {
    for (NCZone *zone in _sublocation.zones) {
        UIBezierPath *zonePath     = [[UIBezierPath alloc] init];
        CAShapeLayer *zoneLayer = [CAShapeLayer layer];
        NSArray *points = zone.points;
        NCLocationPoint *point0 = points[0];
        
        [zonePath moveToPoint:CGPointMake(_imageView.width * point0.x.doubleValue / _sublocation.width,
                                          _imageView.height * (1. - point0.y.doubleValue / _sublocation.height))];
        for (NCLocationPoint *point in zone.points) {
                [zonePath addLineToPoint:CGPointMake(_imageView.width * point.x.doubleValue / _sublocation.width,
                                                     _imageView.height * (1. - point.y.doubleValue / _sublocation.height))];
        }
        [zonePath addLineToPoint:CGPointMake(_imageView.width * point0.x.doubleValue / _sublocation.width,
                                             _imageView.height *(1. - point0.y.doubleValue / _sublocation.height))];
        unsigned int result = 0;
        NSScanner *scanner = [NSScanner scannerWithString:zone.color];
        [scanner setScanLocation:1];
        [scanner scanHexInt:&result];
        
        zoneLayer.hidden = NO;
        zoneLayer.path            = [zonePath CGPath];
        zoneLayer.strokeColor     = [kColorFromHex(result) CGColor];
        zoneLayer.lineWidth       = 2.0;
        zoneLayer.lineJoin        = kCALineJoinRound;
        zoneLayer.fillColor       = [[kColorFromHex(result) colorWithAlphaComponent:0.5] CGColor];
        
        [_imageView.layer addSublayer:zoneLayer];
    }
}

- (IBAction)btnIncreaseFloorPressed:(id)sender {
    if (_floor == _navigineCore.location.sublocations.count - 1)
        return;
    [self setupFloor: ++_floor];
}


- (IBAction)btnDecreaseFloorPressed:(id)sender {
    if (_floor == 0)
        return;
    [self setupFloor: --_floor];
}

# pragma mark NavigineCoreDelegate methods

- (void) navigineCore:(NavigineCore *)navigineCore didUpdateDeviceInfo:(NCDeviceInfo *)deviceInfo {
    NSLog(@"");
}
@end

File: Examples/Swift/Navigine Example/ViewController.swift
//
//  ViewController.swift
//  Swift-New
//
//  Created by Dmitry Stavitsky on 08/07/2018.
//  Copyright © 2018 Navigine. All rights reserved.
//

import UIKit
import Navigine

class ViewController: UIViewController {
  // Constraints
  @IBOutlet weak var imageViewHeightConstraint: NSLayoutConstraint!
  @IBOutlet weak var imageViewWidthConstraint: NSLayoutConstraint!
  
  // UI elements
  @IBOutlet weak var mImageView: UIImageView!
  @IBOutlet weak var mScrollView: UIScrollView!
  @IBOutlet weak var mLblCurrentFloor: UILabel!
  @IBOutlet weak var mBtnStackFloor: UIStackView!
  
  var mEventView = RouteEventView()
  var mErrorView = ErrorView()
  
  // Map elements
  var mPressedPin  = MapPin(venue: NCVenue()) // Selected venue
  var mCurPosition = CurrentLocation() // Current position
  var mFloor = 0
  
  // Path
  var mRoutePath  = UIBezierPath()
  var mRouteLayer = CAShapeLayer()
  var mIsRouting  = false

  // Navigine core attributes
  var mNavigineCore = NavigineCore()
  var mLocation = NCLocation()
  var mSublocation = NCSublocation() // Current sublocation
  
  // Login options
  let userHash   = "0000-0000-0000-0000" // Your user hash
  let serverName = "https://api.navigine.com"
  let locationId = 2872
  
  override func viewDidLoad() {
    super.viewDidLoad()
    mScrollView.delegate = self
    mScrollView.pinchGestureRecognizer?.isEnabled = true
    mScrollView.minimumZoomScale = 1.0
    mScrollView.maximumZoomScale = 2.0
    mScrollView.zoomScale = 1.0
    
    // Create gesture regignizers for long tap and single tap
    let singleTapGesture = UITapGestureRecognizer(target: self, action: #selector(singleTapOnMap))
    let longTapGesture = UILongPressGestureRecognizer(target: self, action: #selector(longTapOnMap))
    mScrollView.addGestureRecognizer(singleTapGesture)
    mScrollView.addGestureRecognizer(longTapGesture)
    
    // Add navigation bar
    mEventView = RouteEventView(frame: CGRect(x: 0.0, y: view.statusBarFrame().size.height, width: view.frame.size.width, height: 80.0))
    view.addSubview(mEventView)
    mEventView.mCancelBtn.addTarget(self, action: #selector(stopRoute), for: .touchUpInside)
    
    // Error view
    let xPos = 0
    let yPos = view.frame.size.height - view.frame.size.height / 8.0
    let width = view.frame.size.width
    let height = view.frame.size.height / 8.0
    mErrorView = ErrorView(frame: CGRect(x: CGFloat(xPos), y: CGFloat(yPos), width: CGFloat(width), height: CGFloat(height)))
    view.addSubview(mErrorView)
    
    // Add current position on map
    mImageView.addSubview(mCurPosition)
    view.bringSubview(toFront: mBtnStackFloor!)
    
    // Add loading progress bar while map is downloading
    let spinnerActivity = MBProgressHUD.showAdded(to: view, animated: true)
    spinnerActivity.mode = .determinateHorizontalBar
    spinnerActivity.label.text = "Loading"
    spinnerActivity.detailsLabel.text = "Please Wait!"
    spinnerActivity.isUserInteractionEnabled = false
    
    // Initialize navigation core
    mNavigineCore = NavigineCore(userHash: userHash, server: serverName)
    mNavigineCore.navigationDelegate = self
    
    mNavigineCore.downloadLocation(byId: locationId,
                                   forceReload: true,
                                   processBlock: {(_ loadProcess: Int) in
                                    print("load process: \(loadProcess)")
                                    spinnerActivity.progress = Float(loadProcess)
                                  },
                                   successBlock: {(_ userInfo:[AnyHashable : Any]?) in
                                    self.mNavigineCore.startNavigine()
                                    self.mNavigineCore.startPushManager()
                                    self.setupFloor(self.mFloor)
                                    spinnerActivity.hide(animated: true)
                                    self.mImageView.isUserInteractionEnabled = true },
                                   fail: {(_ error: Error?) in
                                    print("error: \(error.debugDescription)")})
    // Add beacon generators if needed
    mNavigineCore.addBeaconGenerator("F7826DA6-4FA2-4E98-8024-BC5B71E0893E", major: 65463, minor: 38214, timeout: 50, rssiMin: -100, rssiMax: -70)
    // [_navigineCore addBeaconGenerator: @"F7826DA6-4FA2-4E98-8024-BC5B71E0893E" major: 63714 minor:8737 timeout:50 rssiMin:-100 rssiMax:-x70];
    // [_navigineCore addBeaconGenerator: @"8EEB497E-4928-44C6-9D92-087521A3547C" major: 9001  minor:36 timeout:10 rssiMin:-90 rssiMax:-70];
  }
  
  func setupFloor(_ floor: Int) {
    mImageView.removeFromSuperview()
    removeVenuesFromMap() // Remove venues from map
    removeZonesFromMap() // Remove zones from map
    mScrollView.zoomScale = 1.0 // Reset zoom
    mLocation = mNavigineCore.location
    mSublocation = mNavigineCore.location.sublocations[floor] as! NCSublocation
    mImageView.image = UIImage(data: mSublocation.pngImage)
    mScrollView.addSubview(mImageView)
    mBtnStackFloor.isHidden = mLocation.sublocations.count == 1 // Hide buttons if count of sublocations = 0
    mLblCurrentFloor.text = String(floor)
    let imgSize: CGSize = mImageView.image!.size
    let viewSize: CGSize = view.frame.size
    var scale: CGFloat = 1.0
    if (imgSize.width / imgSize.height) > (viewSize.width / viewSize.height) {
      scale = CGFloat(viewSize.height / imgSize.height)
    } else {
      scale = CGFloat(imgSize.width / imgSize.width)
    }
    mScrollView.contentSize = CGSize(width: imgSize.width * CGFloat(scale),
                                     height: imgSize.height * CGFloat(scale))
    // Add constraints
    imageViewWidthConstraint.constant = imgSize.width * CGFloat(scale)
    imageViewHeightConstraint.constant = imgSize.height * CGFloat(scale)
    
    view.layoutIfNeeded()
    drawZones()
    drawVenues()
  }
}

// MARK: - Button actions
extension ViewController {
  
  @IBAction func singleTapOnMap(gesture: UITapGestureRecognizer) {
    if mPressedPin.isSelected {
      mPressedPin.mPopUp.isHidden = true
    }
  }
  
  // Draw route by long tap
  @IBAction func longTapOnMap(gesture: UITapGestureRecognizer) {
    if gesture.state != .began {
      return
    }
    mImageView.viewWithTag(1)?.removeFromSuperview() // Remove destination pin from map
    mNavigineCore.cancelTarget()
    let touchPtInPx = gesture.location(ofTouch: 0, in: mScrollView) // Touch point in pixels
    let touchPtInM = convertPixelsToMeters(srcX: touchPtInPx.x, srcY: touchPtInPx.y, scale: 1) // Touch point in meters
    let targetPt = NCLocationPoint(location: mLocation.id,
                                   sublocation: mSublocation.id,
                                   x: NSNumber(value: Float(touchPtInM.x)),
                                   y: NSNumber(value: Float(touchPtInM.y)))
    mNavigineCore.addTarget(targetPt)
    // Create destination pin on map
    if let imgMarker = UIImage(named: "elmMapPin") {
      let destinationMarker = UIImageView(image: imgMarker)
      destinationMarker.tag = 1
      destinationMarker.transform = CGAffineTransform(scaleX: 1.0 / mScrollView.zoomScale, y: 1.0 / mScrollView.zoomScale)
      destinationMarker.setCenterX(touchPtInPx.x / mScrollView.zoomScale)
      destinationMarker.setCenterY((touchPtInPx.y - imgMarker.size.height / 2.0) / mScrollView.zoomScale)
      destinationMarker.layer.zPosition = 5.0
      mImageView.addSubview(destinationMarker)
      mIsRouting = true
    }
  }
  
  @objc func stopRoute() {
    mImageView.viewWithTag(1)?.removeFromSuperview() // Remove current pin from map
    mRouteLayer.removeFromSuperlayer()
    mRoutePath.removeAllPoints()
    mNavigineCore.cancelTargets()
    mIsRouting = false
    mEventView.isHidden = true
  }
  
  @IBAction func btnIncreaseFloorPressed() {
    let sublocCnt = mLocation.sublocations.count - 1
    if mFloor == sublocCnt {
      return
    }
    else {
      mFloor += 1
      setupFloor(mFloor)
    }
  }
  
  @IBAction func btnDecreaseFloorPressed() {
    if mFloor == 0 {
      return
    }
    else {
      mFloor -= 1
      setupFloor(mFloor)
    }
  }
  
  @IBAction func mapPinTap(_ pinBtn: MapPin) {
    if pinBtn.isSelected {
      pinBtn.isSelected = false
    } else {
      mPressedPin.mPopUp.isHidden = true // Hide last selected pin
      mPressedPin = pinBtn
      pinBtn.mPopUp.isHidden = false // Show currently selected pin
      pinBtn.mPopUp.setCenterX(pinBtn.centerX())
      pinBtn.isSelected = true
      pinBtn.resizeMapPinWithZoom(mScrollView.zoomScale)
      mImageView.addSubview(pinBtn.mPopUp)
    }
  }
}

// MARK: - Draw functions
extension ViewController {
  
  func drawRoute(withPath path: NCRoutePath, andDistance distance: Float) {
    if distance <= 3.0 { // Check that we are close to the finish point of the route
      stopRoute()
    } else {
      mRouteLayer.removeFromSuperlayer()
      mRoutePath.removeAllPoints()
      mRouteLayer = CAShapeLayer()
      mRoutePath = UIBezierPath()
      for obj in path.points {
        if let curPoint = obj as? NCLocationPoint {
          if curPoint.sublocation != mSublocation.id { // If path between different sublocations
            continue
          }
          else {
            let cgPoint: CGPoint = convertMetersToPixels(srcX: CGFloat(truncating: curPoint.x),
                                                         srcY: CGFloat(truncating: curPoint.y),
                                                         scale: mScrollView.zoomScale)
            if mRoutePath.isEmpty {
              mRoutePath.move(to: cgPoint)
            } else {
              mRoutePath.addLine(to: cgPoint)
            }
          }
        }
      }
      mRouteLayer.isHidden = false
      mRouteLayer.path = mRoutePath.cgPath
      mRouteLayer.strokeColor = UIColor(hex: 0x4aadd4).cgColor
      mRouteLayer.lineWidth = 2.0
      mRouteLayer.lineJoin = kCALineJoinRound
      mRouteLayer.fillColor = UIColor.clear.cgColor
      mImageView.layer.addSublayer(mRouteLayer) // Add route layer on map
      mImageView.bringSubview(toFront: mCurPosition)
    }
  }

  func drawVenues() {
    for obj in mSublocation.venues {
      if let curVenue = obj as? NCVenue {
        let mapPin = MapPin(venue: curVenue)
        let xPt = CGFloat(truncating: curVenue.x)
        let yPt = CGFloat(truncating: curVenue.y)
        mapPin.center = convertMetersToPixels(srcX: xPt, srcY: yPt, scale: 1)
        mapPin.addTarget(self, action: #selector(mapPinTap(_:)), for: .touchUpInside)
        mapPin.sizeToFit()
        mImageView.addSubview(mapPin)
        mScrollView.bringSubview(toFront: mapPin)
      }
    }
  }
  
  func drawZones() {
    for obj in mSublocation.zones {
      if let curZone = obj as? NCZone {
        let zonePath = UIBezierPath()
        let zoneLayer = CAShapeLayer()
        var firstPoint = CGPoint(x: 0, y: 0)
        for obj in curZone.points {
          if let curPoint = obj as? NCLocationPoint {
            let xPt = curPoint.x
            let yPt = curPoint.y
            let cgCurPoint: CGPoint = convertMetersToPixels(srcX: CGFloat(truncating: xPt),
                                                            srcY: CGFloat(truncating: yPt),
                                                            scale: 1)
            if zonePath.isEmpty {
              firstPoint = cgCurPoint
              zonePath.move(to: cgCurPoint)
            } else {
              zonePath.addLine(to: cgCurPoint)
            }
          }
        }
        zonePath.addLine(to: firstPoint) // Add first point again to close path
        let hexColor:UInt32 = stringToHex(srcStr: curZone.color) // Parse zone color
        zoneLayer.name = "Zone"
        zoneLayer.path = zonePath.cgPath
        zoneLayer.strokeColor = UIColor(hex: Int(hexColor)).cgColor
        zoneLayer.lineWidth = 2.0
        zoneLayer.lineJoin = kCALineJoinRound
        zoneLayer.fillColor = UIColor(hex: Int(hexColor)).withAlphaComponent(0.5).cgColor
        mImageView.layer.addSublayer(zoneLayer)
      }
    }
  }
  
  func removeVenuesFromMap() {
    for obj in mImageView.subviews {
      if let pin = obj as? MapPin {
        pin.removeFromSuperview()
        pin.mPopUp.removeFromSuperview()
      }
    }
  }
  
  func removeZonesFromMap() {
    for layer in mImageView.layer.sublayers! { // Unsafe code!!!
      if layer.name == "Zone" {
        layer.removeFromSuperlayer()
      }
    }
  }
}

// MARK: - Helper functions
extension ViewController {
  
  // Convert from pixels to meters
  func convertPixelsToMeters(srcX: CGFloat, srcY: CGFloat, scale: CGFloat) -> CGPoint {
    let dstX = (srcX / (mImageView.width() / scale) * CGFloat(mSublocation.width))
    let dstY = ((1.0 - srcY / (mImageView.height() / scale)) * CGFloat(mSublocation.height))
    return CGPoint(x: dstX, y: dstY)
  }
  
  // Convert from meters to pixels
  func convertMetersToPixels(srcX: CGFloat, srcY: CGFloat, scale: CGFloat) -> CGPoint {
    let dstX = CGFloat(mImageView.width() / scale * srcX / CGFloat(mSublocation.width))
    let dstY = CGFloat(mImageView.height() / scale * (1.0 - srcY / CGFloat(mSublocation.height)))
    return CGPoint(x: dstX, y: dstY)
  }
  
  func stringToHex(srcStr: String) -> uint {
    var hexStr: UInt32 = 0
    let strScanner = Scanner(string: srcStr)
    strScanner.scanLocation = 1
    strScanner.scanHexInt32(&hexStr)
    return hexStr
  }
}

extension ViewController: NavigineCoreNavigationDelegate {
  
  func navigineCore(_ navigineCore: NavigineCore, didUpdate deviceInfo: NCDeviceInfo) {
    if let navError = deviceInfo.error {
      // If some error
      mCurPosition.isHidden = true
      mErrorView.mError = navError as NSError
      mErrorView.isHidden = false
    }
    else {
      mErrorView.isHidden = true
      mCurPosition.isHidden = deviceInfo.sublocation != mSublocation.id // Hide current position pin
      let radScale = mImageView.width() / CGFloat(mSublocation.width)
      mCurPosition.center = convertMetersToPixels(srcX: CGFloat(deviceInfo.x), srcY: CGFloat(deviceInfo.y), scale: mScrollView.zoomScale)
      mCurPosition.mRadius = CGFloat(deviceInfo.r) * radScale
      if mIsRouting {
        if let devicePath:NCRoutePath = deviceInfo.paths?.first as? NCRoutePath {
          let lastPoint = devicePath.points.last as? NCLocationPoint // Last point from route
          mImageView.viewWithTag(1)?.isHidden = lastPoint?.sublocation != mSublocation.id // Hide destination pin
          mEventView.isHidden = deviceInfo.sublocation != mSublocation.id // Hide event bar
          let distance = devicePath.lenght
          if distance < 1 {
            mEventView.setFinishTitle()
          }
          else {
            if let firstEvent = devicePath.events.first as? NCRouteEvent {
              mEventView.mEvent = firstEvent
            }
          }
          drawRoute(withPath: devicePath, andDistance: distance)
        }
      }
    }
  }
  
  func navigineCore(_ navigineCore: NavigineCore, didEnterZone zone: NCZone) {
    print("You are enter in zone ", zone.id)
  }
  
  func navigineCore(_ navigineCore: NavigineCore, didExitZone zone: NCZone) {
    print("You are leave zone ", zone.id)
  }

}

extension ViewController: UIScrollViewDelegate {
  
  func viewForZooming(in scrollView: UIScrollView) -> UIView? {
    return mImageView
  }
  
  func scrollViewDidZoom(_ scrollView: UIScrollView) {
    let currentZoom = mScrollView.zoomScale
    if currentZoom < mScrollView.minimumZoomScale || currentZoom > self.mScrollView.maximumZoomScale {
      return
    }
    else {
      // Stay pins at same sizes while zooming
      for obj in mImageView.subviews {
        if let pin = obj as? MapPin {
          pin.resizeMapPinWithZoom(currentZoom)
        }
      }
      mCurPosition.resizeLocationPinWithZoom(zoom: currentZoom) // Stay current position pin at same sizes while zooming
      // Stay destination marker pin at same sizes while zooming
      if let destMarker:UIImageView = mImageView.viewWithTag(1) as? UIImageView {
        destMarker.transform = CGAffineTransform(scaleX: 1.0 / currentZoom, y: 1.0 / currentZoom)
        destMarker.setCenterX(mRoutePath.currentPoint.x)
        destMarker.setCenterY(mRoutePath.currentPoint.y - ((destMarker.image?.size.height)! / 2) / currentZoom)
      }
    }
    
    // Another way to resize pins accordingly zoom
    /* CGAffineTransform transform = _imageView.transform; // Get current matrix
     CGAffineTransform invertedTransform = CGAffineTransformInvert(transform); // Inverse matrix
     _pressedPin.transform = invertedTransform; // Apply transformation to button
     _pressedPin.popUp.transform = invertedTransform; // Apply transformation to popup */
  }
}
